### PR TITLE
[TOPIC-BLE-LLCP] Integrating Min Number of Used chans into controller.

### PR DESCRIPTION
**Pull request by erbr-ot**
Integrating Min Number of used Channels procedure into controller.

This a work in progress, as testing cannot be completed until Channel Map procedure is completed. The functionality was ported over from legacy code. 

Testing so far was made using a hacked EDTT test, confirming that a periph initiated MNUC procedure, resulted in a ChMapUpd arriving at the Central.

### CODE REVIEW

File: subsys/bluetooth/controller/ll_sw/ull_llcp_common.c [hunk 1]
```diff
@@ -570,7 +572,20 @@ static void rp_comm_send_rsp(struct ll_conn *conn, struct proc_ctx *ctx, uint8_t
 		}
 		break;
 	case PROC_MIN_USED_CHANS:
+		/* Spec says (5.2, Vol.6, Part B, Section 5.1.11):
+		     The procedure has completed when the Link Layer acknowledgment of the
+		     LL_MIN_USED_CHANNELS_IND PDU is sent or received.
+		   In effect, for this procedure, this is equivalent to RX of PDU
+		*/
+		/* So we inititate a chmap update procedure, but only if acting as central, just in case ... */
+		if (!conn->lll.role) { // && ull_conn_lll_phy_active(conn, ctx->data.muc.phys)) {
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_common.c:581]: Check `role` againt `BT_HCI_ROLE_MASTER`

File: tests/bluetooth/bsim_bt/edtt_ble_test_app/common/commands.h [hunk 2]
```diff
@@ -235,7 +235,9 @@ enum commands_t {
 	CMD_GAP_READ_IRK_REQ,
 	CMD_GAP_READ_IRK_RSP,
 	CMD_GAP_ROLE_REQ,
-	CMD_GAP_ROLE_RSP
```
Comment (thoh-ot) [inline tests/bluetooth/bsim_bt/edtt_ble_test_app/common/commands.h:238]: This doesn't really seems to belong in this PR

**Review by thoh-ot — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/hci/hci.c [hunk 3]
```diff
@@ -4037,7 +4050,13 @@ int hci_vendor_cmd_handle_common(uint16_t ocf, struct net_buf *cmd,
 		mesh_cmd_handle(cmd, evt);
 		break;
 #endif /* CONFIG_BT_HCI_MESH_EXT */
-
```
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/hci/hci.c:4040]: yeap

File: subsys/bluetooth/controller/ll_sw/ull_llcp_common.c [hunk 4]
```diff
@@ -570,7 +572,20 @@ static void rp_comm_send_rsp(struct ll_conn *conn, struct proc_ctx *ctx, uint8_t
...
+			uint8_t chmap[5];
+			ull_chan_map_get((uint8_t *const)chmap);
+			uint8_t sts = ull_cp_chan_map_update(conn, chmap);
+			printk("MIN_USED %d, %d, %d, %d, %d - %d\n", chmap[0], chmap[1], chmap[2], chmap[3], chmap[4], sts);
```
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_common.c:585]: why not? Ok, I'll remove

File: subsys/bluetooth/controller/ll_sw/ull_llcp_hci.c [hunk 5]
```diff
@@ -884,6 +884,27 @@ uint8_t ll_enc_req_send(uint16_t handle, uint8_t const *const rand,
 }
 #endif /* CONFIG_BT_CTLR_LE_ENC */
 
+#if defined(CONFIG_BT_CTLR_MIN_USED_CHAN)
+uint8_t ll_set_min_used_chans(uint16_t handle, uint8_t const phys,
+		     uint8_t const min_used_chans)
+{
+	struct ll_conn *conn;
+
+	printk("ll_set_min_used_chans:", phys, min_used_chans);
```
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_hci.c:893]: ok then

File: subsys/bluetooth/controller/ll_sw/ull_llcp_hci.c [hunk 6]
```diff
@@ -884,6 +884,27 @@ uint8_t ll_enc_req_send(uint16_t handle, uint8_t const *const rand,
...
+	conn = ll_connected_get(handle);
+	if (!conn) {
+		return BT_HCI_ERR_UNKNOWN_CONN_ID;
+	}
+
+	if (!conn->lll.role) {
+		return BT_HCI_ERR_CMD_DISALLOWED;
+	}
+	printk(" %d %d\n", phys, min_used_chans);
```
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_hci.c:902]: you just keep at it

File: tests/bluetooth/bsim_bt/edtt_ble_test_app/common/commands.h (hunk 2 above)
Comment (erbr-ot) [reply tests/bluetooth/bsim_bt/edtt_ble_test_app/common/commands.h:238]: Well, no but then again, it is what is needed to have edtt support for set min used chans

Comment (thoh-ot) [reply tests/bluetooth/bsim_bt/edtt_ble_test_app/common/commands.h:238]: Also why is the values set to 255, that doesn't seems to follow any pattern in this enum?

Comment (erbr-ot) [reply tests/bluetooth/bsim_bt/edtt_ble_test_app/common/commands.h]: yeah, you're right - I took it out again. Once I get around to cleaning up the EDTT end of this, I'll add them back and sort out the numbering. The 255 is the next command avail, I think somehow this enum is not entirely up-to-date

File: subsys/bluetooth/controller/ll_sw/ull_llcp_common.c (hunk 1 above)
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_common.c]: yeap

Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_common.c]: and bring back the reminder of the if clause - that was temp commented out...

File: subsys/bluetooth/controller/ll_sw/ull_llcp_common.c [hunk 7]
```diff
@@ -570,7 +572,19 @@ static void rp_comm_send_rsp(struct ll_conn *conn, struct proc_ctx *ctx, uint8_t
...
+		/* Spec says (5.2, Vol.6, Part B, Section 5.1.11):
+		     The procedure has completed when the Link Layer acknowledgment of the
+		     LL_MIN_USED_CHANNELS_IND PDU is sent or received.
+		   In effect, for this procedure, this is equivalent to RX of PDU
+		*/
+		/* So we inititate a chmap update procedure, but only if acting as central, just in case ... */
+		if (conn->lll.role == BT_HCI_ROLE_MASTER && ull_conn_lll_phy_active(conn, ctx->data.muc.phys)) {
+			uint8_t chmap[5];
+			ull_chan_map_get((uint8_t *const)chmap);
+			uint8_t sts = ull_cp_chan_map_update(conn, chmap);
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp_common.c:584]: unused variable `sts`

**Review by thoh-ot — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_llcp_common.c (hunk 7 above)
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_common.c]: yes

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c [hunk 8]
```diff
@@ -459,10 +459,6 @@ uint8_t ull_cp_chan_map_update(struct ll_conn *conn, uint8_t chm[5])
 {
 	struct proc_ctx *ctx;
 
-	if (conn->lll.role != BT_HCI_ROLE_MASTER) {
-		return BT_HCI_ERR_CMD_DISALLOWED;
-	}
-
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp.c:465]: Think this makes test_channel_map_update_sla_loc fail in tests/bluetooth/controller/ctrl_chmu/bluetooth.controller.ctrl_chmu.test

**Review by thoh-ot — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_llcp.c (hunk 8 above)
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.c]: yes

**Review by thoh-ot — APPROVED** [review]

**Review by kruithofa — APPROVED** [review]

**Pull request merged**